cmd/tail.go

```diff
@@ -0,0 +1,137 @@
+package cmd
+
+import (
+	"fmt"
+	"io"
+	"strings"
+	"time"
+
+	"github.com/Benchkram/errz"
+	"github.com/hpcloud/tail"
+	"github.com/puppetlabs/wash/api/client"
+	apitypes "github.com/puppetlabs/wash/api/types"
+	cmdutil "github.com/puppetlabs/wash/cmd/util"
+	"github.com/puppetlabs/wash/config"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+)
+
+func tailCommand() *cobra.Command {
+	tailCmd := &cobra.Command{
+		Use:   "tail -f <file> [<file>...]",
+		Short: "Displays new output of files or resources that support the stream action",
+	}
+
+	tailCmd.Flags().BoolP("follow", "f", false, "Set the output format (json or yaml)")
+	if err := viper.BindPFlag("follow", tailCmd.Flags().Lookup("follow")); err != nil {
+		cmdutil.ErrPrintf("%v\n", err)
+	}
+
+	tailCmd.RunE = toRunE(tailMain)
+
+	return tailCmd
+}
+
+type line struct {
+	tail.Line
+	source string
+}
+
+type chanWriter struct {
+	name string
+	out  chan line
+}
+
+func (w chanWriter) Write(b []byte) (int, error) {
+	s := strings.TrimSuffix(string(b), "\n")
+	w.out <- line{Line: tail.Line{Text: s, Time: time.Now()}, source: w.name}
+	return len(b), nil
+}
+
+func tailMain(cmd *cobra.Command, args []string) exitCode {
+	follow := viper.GetBool("follow")
+	if !follow {
+		cmdutil.ErrPrintf("Please use -f, other operations are not yet supported\n")
+		return exitCode{1}
+	}
+
+	// If no paths are declared, try to stream the current directory/resource
+	if len(args) == 0 {
+		args = []string{"."}
+	}
+
+	conn := client.ForUNIXSocket(config.Socket)
+	agg := make(chan line)
+
+	// Separate paths into streamable resources and files
+	var files []string
+	for _, path := range args {
+		apiPath, err := client.APIKeyFromPath(path)
+		if err != nil {
+			// Not a resource
+			files = append(files, path)
+			continue
+		}
+
+		if stream, err := conn.Stream(apiPath); err == nil {
+			defer func() { errz.Log(stream.Close()) }()
+			// Start copying the stream to the aggregate channel
+			go func(src string) {
+				_, err := io.Copy(chanWriter{name: src, out: agg}, stream)
+				if err != nil {
+					agg <- line{Line: tail.Line{Time: time.Now(), Err: err}, source: src}
+				}
+			}(path)
+		} else {
+			if errObj, ok := err.(*apitypes.ErrorObj); ok && errObj.Kind == "unsupported-action" {
+				// The resource exists but does not support the streaming action, try to read it as a file.
+				// TODO: should we go through the FS or try to use the /fs/read endpoint?
+				files = append(files, path)
+			} else {
+				cmdutil.ErrPrintf("%v\n", err)
+				return exitCode{1}
+			}
+		}
+	}
+
+	// Start at the end of the file
+	tailConf := tail.Config{
+		Follow:   true,
+		Location: &tail.SeekInfo{Offset: 0, Whence: 2},
```

Comment:
What does setting the Location do functionality-wise? A comment would help here.

Reply:
That's exactly what the comment says. I guess I can clarify it to relate more clearly to Location.

Reply:
Oh yeah, brain fart on my part. Maybe a bit more clarification would help in case other people are as mentally lazy as me.